rmw_fastrtps_shared_cpp/src/custom_subscriber_info.cpp

```diff
@@ -21,47 +21,264 @@
+        event_type));
+
+    user_data_[event_type] = nullptr;
+    on_new_event_cb_[event_type] = nullptr;
+  }
+}
+
+void
+SubListener::set_on_new_message_callback(
+  const void * user_data,
+  rmw_event_callback_t callback)
+{
+  std::unique_lock<std::mutex> lock_mutex(on_new_message_m_);
+
+  if (callback) {
+    auto unread_messages = get_unread_messages();
+
+    if (0 < unread_messages) {
+      callback(new_message_user_data_, unread_messages);
+    }
+
+    new_message_user_data_ = user_data;
+    on_new_message_cb_ = callback;
+
+    eprosima::fastdds::dds::StatusMask status_mask =
+      subscriber_info_->data_reader_->get_status_mask();
+    subscriber_info_->data_reader_->set_listener(
+      this,
+      status_mask << eprosima::fastdds::dds::StatusMask::data_available());
+  } else {
+    eprosima::fastdds::dds::StatusMask status_mask =
+      subscriber_info_->data_reader_->get_status_mask();
+    subscriber_info_->data_reader_->set_listener(
+      this,
+      status_mask >> eprosima::fastdds::dds::StatusMask::data_available());
+
+    new_message_user_data_ = nullptr;
+    on_new_message_cb_ = nullptr;
+  }
+}
+
+size_t SubListener::get_unread_messages()
+{
+  return subscriber_info_->data_reader_->get_unread_count();
+}
+
+void
+SubListener::on_data_available(
+  eprosima::fastdds::dds::DataReader *)
+{
+  std::unique_lock<std::mutex> lock_mutex(on_new_message_m_);
+
+  if (on_new_message_cb_) {
+    auto unread_messages = get_unread_messages();
+
+    if (0 < unread_messages) {
+      on_new_message_cb_(new_message_user_data_, unread_messages);
+    }
+  }
+}
+
 void
 SubListener::on_requested_deadline_missed(
-  eprosima::fastdds::dds::DataReader * /* reader */,
+  eprosima::fastdds::dds::DataReader *,
   const eprosima::fastdds::dds::RequestedDeadlineMissedStatus & status)
 {
-  std::lock_guard<std::mutex> lock(internalMutex_);
-
-  // the change to liveliness_lost_count_ needs to be mutually exclusive with
-  // rmw_wait() which checks hasEvent() and decides if wait() needs to be called
-  ConditionalScopedLock clock(conditionMutex_, conditionVariable_);
+  std::unique_lock<std::mutex> lock_mutex(on_new_event_m_);
 
   // Assign absolute values
   requested_deadline_missed_status_.total_count = status.total_count;
   // Accumulate deltas
   requested_deadline_missed_status_.total_count_change += status.total_count_change;
 
-  deadline_changes_.store(true, std::memory_order_relaxed);
+  deadline_changes_ = true;
 
-  std::unique_lock<std::mutex> lock_mutex(on_new_event_m_);
-
-  if (on_new_event_cb_) {
-    on_new_event_cb_(user_data_, 1);
-  } else {
-    unread_events_count_++;
+  if (on_new_event_cb_[RMW_EVENT_REQUESTED_DEADLINE_MISSED]) {
+    on_new_event_cb_[RMW_EVENT_REQUESTED_DEADLINE_MISSED](user_data_[
+        RMW_EVENT_REQUESTED_DEADLINE_MISSED], 1);
   }
+
+  event_guard[RMW_EVENT_REQUESTED_DEADLINE_MISSED].set_trigger_value(true);
```

Comment:
This code is repeated several times with different event types.
I think it is worth having a `trigger_event(rmw_event_type_t event_type)` method.

Reply:
Done in 68aacaa